# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [46]:
import os
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
%matplotlib inline
pd.options.display.float_format = '{:.2f}'.format #breaking out of scientific notations for the full notebook
os.getcwd()

'C:\\Users\\Vatsla Issar\\Downloads'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [47]:
df=pd.read_csv('step3_output.csv')
df.head(3)

Unnamed: 0                           Name       state  summit_elev  \
0           2               Hilltop Ski Area      Alaska         2090   
1           4            Sunrise Park Resort     Arizona        11100   
2           5  Yosemite Ski & Snowboard Area  California         7800   

   vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  ...  \
0            294      0       0.00          0          0     0  ...   
1           1800      0       0.00          0          1     2  ...   
2            600      0       0.00          0          0     0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0              30.00           30.00            150.00      36.00   
1             800.00           80.00            115.00      49.00   
2              88.00          174.87            110.00      84.00   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            69.00         30.00         34.00             152.00   
1           250.00         74.00         78.00             104.00   
2           300.00         47.00         47.00             107.00   

   NightSkiing_ac  clusters  
0           30.00         2  
1           80.00         1  
2          100.40         1  

[3 rows x 27 columns]

In [48]:
df=df.drop(['Unnamed: 0'], axis=1)
df.head(3)

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0       0.00          0          0     0       1  ...   
1      0       0.00          0          1     2       3  ...   
2      0       0.00          0          0     0       1  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0              30.00           30.00            150.00      36.00   
1             800.00           80.00            115.00      49.00   
2              88.00          174.87            110.00      84.00   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            69.00         30.00         34.00             152.00   
1           250.00         74.00         78.00             104.00   
2           300.00         47.00         47.00             107.00   

   NightSkiing_ac  clusters  
0           30.00         2  
1           80.00         1  
2          100.40         1  

[3 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [49]:
state_d=pd.get_dummies(df['state'])
state_d.head()

Alaska  Arizona  California  Colorado  Connecticut  Idaho  Illinois  \
0       1        0           0         0            0      0         0   
1       0        1           0         0            0      0         0   
2       0        0           1         0            0      0         0   
3       0        0           1         0            0      0         0   
4       0        0           1         0            0      0         0   

   Indiana  Iowa  Maine  ...  Rhode Island  South Dakota  Tennessee  Utah  \
0        0     0      0  ...             0             0          0     0   
1        0     0      0  ...             0             0          0     0   
2        0     0      0  ...             0             0          0     0   
3        0     0      0  ...             0             0          0     0   
4        0     0      0  ...             0             0          0     0   

   Vermont  Virginia  Washington  West Virginia  Wisconsin  Wyoming  
0        0         0           0              0          0        0  
1        0         0           0              0          0        0  
2        0         0           0              0          0        0  
3        0         0           0              0          0        0  
4        0         0           0              0          0        0  

[5 rows x 35 columns]

In [50]:
df = pd.concat([df, state_d], axis=1)

In [51]:
df=df.drop('state', axis=1)

In [52]:
df.head().T #state dropped and state_d (dummies of state) are included.

0                    1  \
Name               Hilltop Ski Area  Sunrise Park Resort   
summit_elev                    2090                11100   
vertical_drop                   294                 1800   
trams                             0                    0   
fastEight                      0.00                 0.00   
fastSixes                         0                    0   
fastQuads                         0                    1   
quad                              0                    2   
triple                            1                    3   
double                            0                    1   
surface                           2                    0   
total_chairs                      3                    7   
Runs                          13.00                65.00   
TerrainParks                   1.00                 2.00   
LongestRun_mi                  1.00                 1.20   
SkiableTerrain_ac             30.00               800.00   
Snow Making_ac                30.00                80.00   
daysOpenLastYear             150.00               115.00   
yearsOpen                     36.00                49.00   
averageSnowfall               69.00               250.00   
AdultWeekday                  30.00                74.00   
AdultWeekend                  34.00                78.00   
projectedDaysOpen            152.00               104.00   
NightSkiing_ac                30.00                80.00   
clusters                          2                    1   
Alaska                            1                    0   
Arizona                           0                    1   
California                        0                    0   
Colorado                          0                    0   
Connecticut                       0                    0   
Idaho                             0                    0   
Illinois                          0                    0   
Indiana                           0                    0   
Iowa                              0                    0   
Maine                             0                    0   
Maryland                          0                    0   
Massachusetts                     0                    0   
Michigan                          0                    0   
Minnesota                         0                    0   
Missouri                          0                    0   
Montana                           0                    0   
Nevada                            0                    0   
New Hampshire                     0                    0   
New Jersey                        0                    0   
New Mexico                        0                    0   
New York                          0                    0   
North Carolina                    0                    0   
Ohio                              0                    0   
Oregon                            0                    0   
Pennsylvania                      0                    0   
Rhode Island                      0                    0   
South Dakota                      0                    0   
Tennessee                         0                    0   
Utah                              0                    0   
Vermont                           0                    0   
Virginia                          0                    0   
Washington                        0                    0   
West Virginia                     0                    0   
Wisconsin                         0                    0   
Wyoming                           0                    0   

                                               2                 3  \
Name               Yosemite Ski & Snowboard Area  Donner Ski Ranch   
summit_elev                                 7800              8012   
vertical_drop                                600               750   
trams                                          0                 0   
fastEight              

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [53]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [54]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [55]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [56]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred=lm.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [57]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score, mean_absolute_error

In [58]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test,y_pred))

-1.8416758824177747e+21
132277106080.53162


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [59]:
lm.intercept_

33670536148.367386

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [60]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
a=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

In [82]:
sort_a=a.sort_values(by='Coefficient', ascending=False)
sort_a

Coefficient
fastEight         15111595917398.34
fastSixes          2266707211617.41
total_chairs       1487428501090.50
trams              1181721441135.81
surface             851124018754.76
New York            844410111382.01
double              818117601232.87
Michigan            798526559807.46
triple              704510790612.23
New Hampshire       692874869387.01
Wisconsin           662771699728.64
Pennsylvania        662771699727.31
Minnesota           596613294222.68
California          596613294222.26
Vermont             559855140543.76
Idaho               519962824333.43
New Mexico          519962824332.63
Wyoming             476149245218.55
Maine               476149245217.51
Colorado            476149245217.13
Connecticut         476149245216.71
Ohio                427209636474.69
Washington          427209636473.40
Massachusetts       427209636472.17
quad                424748588572.54
Iowa                371121608117.18
Montana             371121608116.86
Oregon              371121608116.60
Illinois            371121608116.54
Utah                371121608116.52
West Virginia       371121608114.64
North Carolina      371121608113.29
fastQuads           338474962011.39
Indiana             303953330738.32
Missouri            303953330737.92
Virginia            303953330736.01
Alaska              215585739934.33
South Dakota        215585739934.24
Arizona             215585739933.92
Rhode Island        215585739933.61
Nevada              215585739933.58
Tennessee           215585739932.12
Maryland                21752839.65
New Jersey                 18768.77
AdultWeekday                  10.63
summit_elev                    1.77
vertical_drop                  1.59
clusters                       1.43
daysOpenLastYear               1.25
LongestRun_mi                  1.14
projectedDaysOpen              1.05
Snow Making_ac                 0.84
averageSnowfall                0.76
Runs                           0.67
TerrainParks                   0.54
SkiableTerrain_ac              0.34
yearsOpen                      0.17
NightSkiing_ac                 0.15

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [83]:
df2=pd.read_csv('step3_output.csv') #Loaded again because we created dummies for state earlier (multiple levels)
df2=df2.drop(['Unnamed: 0'], axis=1) #df=df.drop(state_d,axis=1) would have worked but i wanted to keep it seperate
df2.head().T

0                    1  \
Name               Hilltop Ski Area  Sunrise Park Resort   
state                        Alaska              Arizona   
summit_elev                    2090                11100   
vertical_drop                   294                 1800   
trams                             0                    0   
fastEight                      0.00                 0.00   
fastSixes                         0                    0   
fastQuads                         0                    1   
quad                              0                    2   
triple                            1                    3   
double                            0                    1   
surface                           2                    0   
total_chairs                      3                    7   
Runs                          13.00                65.00   
TerrainParks                   1.00                 2.00   
LongestRun_mi                  1.00                 1.20   
SkiableTerrain_ac             30.00               800.00   
Snow Making_ac                30.00                80.00   
daysOpenLastYear             150.00               115.00   
yearsOpen                     36.00                49.00   
averageSnowfall               69.00               250.00   
AdultWeekday                  30.00                74.00   
AdultWeekend                  34.00                78.00   
projectedDaysOpen            152.00               104.00   
NightSkiing_ac                30.00                80.00   
clusters                          2                    1   

                                               2                 3  \
Name               Yosemite Ski & Snowboard Area  Donner Ski Ranch   
state                                 California        California   
summit_elev                                 7800              8012   
vertical_drop                                600               750   
trams                                          0                 0   
fastEight                                   0.00              0.00   
fastSixes                                      0                 0   
fastQuads                                      0                 0   
quad                                           0                 0   
triple                                         1                 1   
double                                         3                 5   
surface                                        1                 2   
total_chairs                                   5                 8   
Runs                                       10.00             52.00   
TerrainParks                                2.00              2.00   
LongestRun_mi                               0.40              1.50   
SkiableTerrain_ac                          88.00            505.00   
Snow Making_ac                            174.87             60.00   
daysOpenLastYear                          110.00            163.00   
yearsOpen                                  84.00             82.00   
averageSnowfall                           300.00            400.00   
AdultWeekday                               47.00             75.00   
AdultWeekend                               47.00             75.00   
projectedDaysOpen                         107.00            170.00   
NightSkiing_ac                            100.40            100.40   
clusters                                       1                 1   

                               4  
Name               June Mountain  
state                 California  
summit_elev                10090  
vertical_drop               2590  
trams                          0  
fastEight                   0.00  
fastSixes                      0  
fastQuads                      2  
quad                           0  
triple                         0  
double                         4  
surface                        1  
total_chairs                   7  
Runs                       41.00

In [84]:
df2=df2.drop('state', axis=1)
df2.head().T

0                    1  \
Name               Hilltop Ski Area  Sunrise Park Resort   
summit_elev                    2090                11100   
vertical_drop                   294                 1800   
trams                             0                    0   
fastEight                      0.00                 0.00   
fastSixes                         0                    0   
fastQuads                         0                    1   
quad                              0                    2   
triple                            1                    3   
double                            0                    1   
surface                           2                    0   
total_chairs                      3                    7   
Runs                          13.00                65.00   
TerrainParks                   1.00                 2.00   
LongestRun_mi                  1.00                 1.20   
SkiableTerrain_ac             30.00               800.00   
Snow Making_ac                30.00                80.00   
daysOpenLastYear             150.00               115.00   
yearsOpen                     36.00                49.00   
averageSnowfall               69.00               250.00   
AdultWeekday                  30.00                74.00   
AdultWeekend                  34.00                78.00   
projectedDaysOpen            152.00               104.00   
NightSkiing_ac                30.00                80.00   
clusters                          2                    1   

                                               2                 3  \
Name               Yosemite Ski & Snowboard Area  Donner Ski Ranch   
summit_elev                                 7800              8012   
vertical_drop                                600               750   
trams                                          0                 0   
fastEight                                   0.00              0.00   
fastSixes                                      0                 0   
fastQuads                                      0                 0   
quad                                           0                 0   
triple                                         1                 1   
double                                         3                 5   
surface                                        1                 2   
total_chairs                                   5                 8   
Runs                                       10.00             52.00   
TerrainParks                                2.00              2.00   
LongestRun_mi                               0.40              1.50   
SkiableTerrain_ac                          88.00            505.00   
Snow Making_ac                            174.87             60.00   
daysOpenLastYear                          110.00            163.00   
yearsOpen                                  84.00             82.00   
averageSnowfall                           300.00            400.00   
AdultWeekday                               47.00             75.00   
AdultWeekend                               47.00             75.00   
projectedDaysOpen                         107.00            170.00   
NightSkiing_ac                            100.40            100.40   
clusters                                       1                 1   

                               4  
Name               June Mountain  
summit_elev                10090  
vertical_drop               2590  
trams                          0  
fastEight                   0.00  
fastSixes                      0  
fastQuads                      2  
quad                           0  
triple                         0  
double                         4  
surface                        1  
total_chairs                   7  
Runs                       41.00  
TerrainParks                2.00  
LongestRun_mi               2.00  
SkiableTerrain_ac        1500.00  
Snow Making_ac            174.87  
daysOpenLastYear      

In [85]:
X2 = df2.drop(['Name','AdultWeekend'], axis=1)
y2 = df2['AdultWeekend'] 

In [86]:
scaler2 = preprocessing.StandardScaler().fit(X2)

In [87]:
X_scaled2=scaler2.transform(X2)

In [88]:
y2 = y2.ravel()

X_train2, X_test2, y_train2, y_test2 = train_test_split(X_scaled2, y2, test_size=0.25, random_state=1)

In [89]:
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X_train2,y_train2)

In [90]:
y_pred2=lm2.predict(X_test2)

In [91]:
print(explained_variance_score(y_test2, y_pred2))
print(mean_absolute_error(y_test2,y_pred2))

0.6193833449199353
6.539623766353013


In [92]:
lm2.intercept_

56.36473440725649

In [93]:
b=pd.DataFrame(abs(lm2.coef_), X2.columns, columns=['Coefficient'])
b.sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday             11.68
summit_elev               3.20
daysOpenLastYear          2.43
vertical_drop             1.89
clusters                  1.80
averageSnowfall           1.73
surface                   1.29
projectedDaysOpen         1.29
quad                      1.21
double                    0.83
triple                    0.70
Runs                      0.53
TerrainParks              0.53
Snow Making_ac            0.47
yearsOpen                 0.44
total_chairs              0.38
SkiableTerrain_ac         0.31
NightSkiing_ac            0.22
LongestRun_mi             0.12
fastQuads                 0.04
fastEight                 0.00
fastSixes                 0.00
trams                     0.00

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [94]:
'base_elev' in df2.columns #base_elev not appearing because it was dropped in step3 as it was highly collinear with 4 features

False

In [74]:
df3=df2.drop(['summit_elev'],axis=1)

In [75]:
df3.columns

Index(['Name', 'vertical_drop', 'trams', 'fastEight', 'fastSixes', 'fastQuads',
       'quad', 'triple', 'double', 'surface', 'total_chairs', 'Runs',
       'TerrainParks', 'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters'],
      dtype='object')

In [76]:
X3 = df3.drop(['Name','AdultWeekend'], axis=1)
y3 = df3['AdultWeekend'] 

In [77]:
scaler3 = preprocessing.StandardScaler().fit(X3)

In [78]:
X_scaled3=scaler3.transform(X3)

In [79]:
y3 = y3.ravel()

X_train3, X_test3, y_train3, y_test3 = train_test_split(X_scaled3, y3, test_size=0.25, random_state=1)

In [80]:
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X_train3,y_train3)
y_pred3=lm3.predict(X_test3)
print(explained_variance_score(y_test3, y_pred3))
print(mean_absolute_error(y_test3,y_pred3))
print(lm3.intercept_)

0.5896428903379245
6.845906805508754
56.36895795894507


In [95]:
c=pd.DataFrame(abs(lm3.coef_), X3.columns, columns=['Coefficient'])
c.sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday             11.40
averageSnowfall           2.59
vertical_drop             2.30
daysOpenLastYear          1.96
SkiableTerrain_ac         1.69
clusters                  1.11
double                    1.10
quad                      1.06
surface                   0.99
projectedDaysOpen         0.84
TerrainParks              0.83
total_chairs              0.64
triple                    0.61
yearsOpen                 0.27
Snow Making_ac            0.21
NightSkiing_ac            0.12
fastQuads                 0.06
Runs                      0.06
LongestRun_mi             0.05
trams                     0.00
fastSixes                 0.00
fastEight                 0.00

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -1.8416e+21 | 132277106080.53 |-|
| Model 2. | 0.61938| 6.5396 |'state'|
| Model 3. | 0.58964 | 6.8459 |'state','summit_elev'|

Model Selection:

In [ ]:
#Explained variance is the maximum for Model 2 (after dropping state). But model 3 makes sense too because collinear features
#have been dropped. 
#But assuming that explained variance as the only criteria, I would pick Model2.